broker/broker_test.go

```diff
@@ -0,0 +1,111 @@
+package broker
+
+import (
+	"sync"
+	"testing"
+)
+
+func TestNewSupervisor(t *testing.T) {
+	supervisor := NewSupervisor()
+	if supervisor.clientsMapping == nil {
+		t.Fatalf("error, expected clientsMapping to be a nil pointer, instead got %v", supervisor.clientsMapping)
+	}
+	if supervisor.clientsMutex == nil {
+		t.Fatalf("error, expected clientsMutex to be a nil pointer, instead got %v", supervisor.clientsMutex)
+	}
+	if supervisor.IsRunning != false {
+		t.Fatalf("error, expected IsRunning to be a false, instead got %v", supervisor.IsRunning)
+	}
+	if supervisor.clientOperationChannel == nil {
+		t.Fatalf("error, expected clientOperationChannel to be a nil pointer, instead got %v", supervisor.clientOperationChannel)
+	}
+	if supervisor.stoppedSignal == nil {
+		t.Fatalf("error, expected stoppedSignal to be a nil pointer, instead got %v", supervisor.stoppedSignal)
+	}
+	if supervisor.startedSignal == nil {
+		t.Fatalf("error, expected startedSignal to be a nil pointer, instead got %v", supervisor.startedSignal)
+	}
+	if supervisor.mainWg == nil {
+		t.Fatalf("error, expected mainWg to be a nil pointer, instead got %v", supervisor.mainWg)
+	}
+	if supervisor.clientCount != 0 {
+		t.Fatalf("error, expected clientCount to be a zero, instead got %v", supervisor.clientCount)
+	}
+}
+
+func TestRun(t *testing.T) {
+	var wg sync.WaitGroup
+	supervisor := NewSupervisor()
+	wg.Add(1)
+	go func() {
+		supervisor.START()
+		wg.Done()
+	}()
+	supervisor.WaitUntilReady()
+	if supervisor.IsRunning != true {
+		t.Fatalf("error, expected IsRunnint to be True, instead got %v", supervisor.IsRunning)
+	}
+	supervisor.STOP()
+	wg.Wait()
```

Comment:
question: Can you explain to me why we are using go routine and "wait-group" instead of something like this: 
```
        supervisor.START()
		
	
	supervisor.WaitUntilReady()
	if supervisor.IsRunning != true {
		t.Fatalf("error, expected IsRunnint to be True, instead got %v", supervisor.IsRunning)
	}
	supervisor.STOP()
	
```

Reply:
and the IsRunning should not be true ?

Reply:
For testing, we are running the supervisor in a separate "thread".
The parent function (the main, if you wish) is responsible to assert things only when things are supposed to be in certain way.
So, `WaitUntilReady` is expected to unblock only after the supervisor is healthy and ready to make some work.
Then we stop the supervisor.
Note that the `wg.Done()` is made just after the `supervisor.Start()` returns. If we weren't using the wait group, we might try to assert IsRunning == false before the supervisor really stopped.
So, in brief, we are using the wait group to make sure we are making the last assert only after we have guarantees about the supervisor have finished.